src/Broadway/EventSourcing/DomainMessageTypeEnrichment/DomainMessageTypeEventStreamDecorator.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace Broadway\EventSourcing\DomainMessageTypeEnrichment;
+
+use Broadway\Domain\DomainEventStream;
+use Broadway\Domain\DomainEventStreamInterface;
+use Broadway\EventSourcing\EventStreamDecoratorInterface;
+
+class DomainMessageTypeEventStreamDecorator implements EventStreamDecoratorInterface
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function decorateForWrite($aggregateType, $aggregateIdentifier, DomainEventStreamInterface $eventStream)
+    {
+        $messages = array();
+
+        foreach ($eventStream as $message) {
+            $payload = $message->getPayload();
+            if (method_exists($payload, 'getType')) {
```

Comment:
That's more less the plan. A new EventInterface could exist with the getType or messageName method defined. The decorator would use that to create new DomainMessage with that new message key.